### PR TITLE
Add new merge_variables lookup plugin

### PR DESCRIPTION
**Pull request by rlenferink**
##### SUMMARY

This adds a new lookup plugin which can be used to merge lists or dicts. This to be able to merge variables that are specified in both the group_vars and the host_vars for usage in tasks. This is only for chosen variables so that the [DEFAULT_HASH_BEHAVIOR](https://docs.ansible.com/ansible/latest/reference_appendices/config.html#default-hash-behaviour) can be kept the default (`replace`).

For example in the group_vars a list of packages to install can be specified where on some nodes you want to extend this list with some additional packages. Normally if a variable is specified in the host_vars it overrides the group_vars, but the idea behind this lookup plugin is that it is possible to retain both the group_vars as the host_vars (only for the specified variable).

Example:

```yaml
# group_vars/node_collection/packages.yml
---
nodes__packages:
- package1
- package2
- package3

# host_vars/node2/packages.yml
node2__packages:
- extra_package1
```

In the role using this variable (e.g. `_packages_definition: "{{ lookup('merge_variables', '__packages') }}"`, the lookup plugin is used to merge the variables together and execute the tasks (in this example installing a package).

In the above scenario all nodes part of the `node_collection` group will have `package[1,2,3]` installed, where `node2` has `package[1,2,3]` and `extra_package1` installed.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
merge_variables lookup plugin

##### ADDITIONAL INFORMATION

The same behaviour can be achieved using a combination of other functionality, e.g.

```yaml
merged_list: "{{ lookup('vars', *query('varnames', '^.+__merge_list$')) | flatten(levels=1) }}"
merged_list_with_initial_value: "{{ (lookup('vars', *query('varnames', '^.+__merge_list$')) + [testlist_initial_value]) | flatten(levels=1) }}"

merged_dict: "{{ lookup('vars', *query('varnames', '^.+__merge_dict$')) | combine({}, recursive=True, list_merge='append') }}"
merged_dict_with_initial_value: "{{ lookup('vars', *query('varnames', '^.+__merge_dict$')) | combine(testdict_initial_value, recursive=True, list_merge='append') }}"
```

This, however, is quite a long and error-prone sentence when being repeated in different roles/group_vars/host_vars.

Having this available in a separate lookup plugin would be:

```yaml
merged_list: "{{ lookup('merge_variables', '__merge_list') }}"
merged_list_with_initial_value: "{{ lookup('merge_variables', '__merge_list', initial_value=testlist_initial_value) }}"

merged_dict: "{{ lookup('merge_variables', '__merge_dict') }}"
merged_dict_with_initial_value: "{{ lookup('merge_variables', '__merge_dict', initial_value=testdict_initial_value) }}"
```

/cc @m-a-r-k-e

### CODE REVIEW

File: changelogs/fragments/5533-add-merge-variables-lookup-plugin.yml
Comment (felixfontein) [inline changelogs/fragments/5533-add-merge-variables-lookup-plugin.yml:2]: New plugins do not need a changelog fragment. What is needed is an entry in `.github/BOTMETA.yml` though.

File: tests/integration/targets/lookup_merge_variables/runme.sh
Comment (felixfontein) [inline tests/integration/targets/lookup_merge_variables/runme.sh:9]: Why have a `runme.sh` if you don't do anything special (like installing Python dependencies)?

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:32]: Better have a `override` option that can take a choice of `ignore`, `warn`, `error`, instead of having two different boolean toggles to control this behavior.

Also I would set the default to `error` and add a warning that setting it to anything else will result in undefined behavior since the merge order can change between different playbook runs.

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:60]: This example will not work, since you do not use the FQCN of the plugin.
```suggestion
example_a: "{{ lookup('community.general.merge_variables', '__test_dict') }}"
```
Same below.

**Review by felixfontein — COMMENTED** [review]
This looks like a highly specialized plugin. I can see it can be useful in *some* cases, but how about making it at least a bit more general by allowing to specify a regex instead of a suffix for the variables to merge, as you can for the `vars` lookup? Like: `"{{ lookup('merge_variables', '^.+__packages$') }}"`.

File: tests/integration/targets/lookup_merge_variables/runme.sh
Comment (rlenferink) [reply tests/integration/targets/lookup_merge_variables/runme.sh:9]: Because I want to set the `log_path`, however, the integration tests don't seem to be picking up the `ansible.cfg` I initially created. With the `runme.sh` and using the environment variable that works (I had a look at how other integration tests where passing options, and the `runme.sh` seemed like the way to go).

File: changelogs/fragments/5533-add-merge-variables-lookup-plugin.yml
Comment (rlenferink) [reply changelogs/fragments/5533-add-merge-variables-lookup-plugin.yml:2]: Changelog removed and `BOTMETA.yml` updated

File: plugins/lookup/merge_variables.py
Comment (rlenferink) [reply plugins/lookup/merge_variables.py:60]: Done

File: plugins/lookup/merge_variables.py
Comment (rlenferink) [reply plugins/lookup/merge_variables.py]: Thanks for the feedback, I combined the options and also am giving a warning in case another option than `error` is chosen.

**Comment by rlenferink** [comment]
> This looks like a highly specialized plugin. I can see it can be useful in _some_ cases, but how about making it at least a bit more general by allowing to specify a regex instead of a suffix for the variables to merge, as you can for the `vars` lookup? Like: `"{{ lookup('merge_variables', '^.+__packages$') }}"`.

Thanks for the feedback! This indeed is a specialized plugin that will probably only be used in specific cases. However, I am aware of different users already, and that's why we (Thales) want to contribute this to the Ansible community.

The idea of this is to keep the usage very simple for the users. Based on your feedback, I added a `pattern_type` option for users to select whether they want to search for variables using a `prefix`, `suffix` or `regex`.

Is this something that would look okay to you?

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:41]:
```suggestion
        default: 'error'
        choices:
          - error
          - warn
          - ignore
```

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:39]: Since that information is already provided by the below metadata, why not add some more information instead?
```suggestion
          - The default behavior C(error) makes the plugin fail when a key would be overwritten.
          - When C(warn) and C(ignore) are used, note that it is important to know that the variables
            are sorted by name before being merged. Keys for later variables in this order will overwrite
            keys of the same name for variables earlier in this order. To avoid potential confusion,
            better use I(override=error) whenever possible.
```

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:29]:
```suggestion
        default: 'suffix'
        choices:
          - prefix
          - suffix
          - regex
```

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:27]:
```suggestion
```

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:29]: How about using `regex` as the default? I think this is the most versatile and this would make the plugin behave similar to the `varnames` plugin by default.

File: plugins/lookup/merge_variables.py
Comment (rlenferink) [reply plugins/lookup/merge_variables.py:41]: I didn't know about this `choices`; thanks for the example!

File: plugins/lookup/merge_variables.py
Comment (rlenferink) [reply plugins/lookup/merge_variables.py:39]: Adopted your suggestion

File: plugins/lookup/merge_variables.py
Comment (rlenferink) [reply plugins/lookup/merge_variables.py:29]: Done

Comment (rlenferink) [reply plugins/lookup/merge_variables.py:29]: Initially I was thinking; this should be kept simple. But after learning that the behavior can be overwritten for every lookup call by setting the `ANSIBLE_MERGE_VARIABLES_PATTERN_TYPE` environment variable, I am okay with changing the default to at least have it consistent with the `varnames` plugin. :+1:

Comment (rlenferink) [reply plugins/lookup/merge_variables.py]: Updated it!

**Comment by rlenferink** [comment]
The test failure seems to be because of a timeout:

```
05:04   ERROR: Could not install packages due to an EnvironmentError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Max retries exceeded with url: /packages/b6/40/353c051f77ee5618adaf1fd96f4f6bae9714ed0a22c7142c01c24eb77fe4/setuptools-65.5.1-py3-none-any.whl (Caused by ConnectTimeoutError(<urllib3.connection.VerifiedHTTPSConnection object at 0x7f067e156700>, 'Connection to files.pythonhosted.org timed out. (connect timeout=15)'))
```

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:54]: You could additionally allow to configure this (and the other options that can be configured with env variables) with an Ansible variable.

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:37]:
```suggestion
          - An initial value to start with.
```

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:38]: `required: false` should be omitted since `false` is the default for `required`. But you should always specify a type. I assume you want to allow any kind of input here?
```suggestion
        type: raw
```

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:22]:
```suggestion
          - Depending on the value of I(pattern_type), this is a list of prefixes, suffixes, or regular expressions
            that will be used to match all variables that should be merged.
```

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:23]:
```suggestion
        required: true
        type: list
        elements: str
```

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:16]:
```suggestion
        - This lookup returns the merged result of all variables in scope that match the given prefixes, suffixes, or regular expessions, optionally
```

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:97]: Please add `RETURN = ...` to document the return value.

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:131]: You probably want to check whether `term` is actually a string before calling `self._merge_vars()`.

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:127]: I can imagine that this warning can be very annoying, especially since you cannot disable it.

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:149]:
```suggestion
        display.vvv("The following variables will be merged: {0}".format(var_merge_names))
```
I would use a higher verbosity level, at least `vvv`, maybe even `vvvv`. Same for the output above.

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:155]:
```suggestion
```
Values can potentially contain confidental data that suddenly ends up in logfiles.

File: plugins/lookup/merge_variables.py
Comment (rlenferink) [reply plugins/lookup/merge_variables.py:54]: I can't really find something on how to do this. Do you know where this is documented?

I also don't really need to set it via an Ansible variable, what's you opinion here? I would also be okay with adding this at a later stadium. It is easier to extend with options than to remove options, since that would be a breaking change.

File: plugins/lookup/merge_variables.py
Comment (rlenferink) [reply plugins/lookup/merge_variables.py:38]: Removed the required.

About the type: I actually think that this should be `type: str`, since it allows a prefix, suffix or regex, which are are string values.

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [reply plugins/lookup/merge_variables.py:54]: > I can't really find something on how to do this. Do you know where this is documented?

I'm not sure where it is documented, but there are numerous examples in this collection, for example: https://github.com/ansible-collections/community.general/blob/main/plugins/lookup/passwordstore.py#L121-L125 This plugin option also can be configured in ansible.cfg (`ini:`).

> I also don't really need to set it via an Ansible variable, what's you opinion here? I would also be okay with adding this at a later stadium. It is easier to extend with options than to remove options, since that would be a breaking change.

Some folks prefer environment variables (you apparently do), others prefer Ansible variables (like me). Offering both is usually best so everyone can use whatever they prefer.

File: plugins/lookup/merge_variables.py
Comment (rlenferink) [reply plugins/lookup/merge_variables.py:127]: I added this because I understood this was desirable from comment https://github.com/ansible-collections/community.general/pull/5533#discussion_r1020536690 , but that might also be me having misunderstood that comment.

I am more than happy to remove this, because the plugin description also describes this now.

What do you suggest? Let's remove this?

File: plugins/lookup/merge_variables.py
Comment (rlenferink) [reply plugins/lookup/merge_variables.py:54]: > Some folks prefer environment variables (you apparently do), others prefer Ansible variables (like me). Offering both is usually best so everyone can use whatever they prefer.

Looking at the `passwordstore.py` example, this wasn't too hard to configure. I just added the options.

File: plugins/lookup/merge_variables.py
Comment (rlenferink) [reply plugins/lookup/merge_variables.py:127]: I at least changed this to the `v` level, since apparently I already displayed a warning in case a key will be overwritten.

Comment (rlenferink) [reply plugins/lookup/merge_variables.py:127]: Re-running the integration tests, I just removed this, since I already display a warning in case a key will be overwritten.

**Comment by russoz** [comment]
Hi @rlenferink thanks for the contribution.

A couple of comments:
* It looks good from the code perspective, and tests seem to be alright. I haven't review at nit-picking level, but it looks solid.
* It might be interesting (not necessarily in this PR but as a future enhancement) to refactor out the actual _merging_ code to module_utils, so that it can be reused by other plugins.
* There is another PR with a similar proposition, #4799 , but it's been stuck for a while and the author has just let us know they won't be able to work on it before January. Nonetheless, it might be worthy to take a look at that PR as well, so that you guys don't rewrite the same thing twice, nor step on each other's toes.

**Comment by russoz** [comment]
@rlenferink all good, I have been away for the last month or so myself. Well, let's wait a bit longer and see how it goes. If there is any opportunity not to repeat ourselves, we should take it.

**Comment by felixfontein** [comment]
@rlenferink @russoz @indelog ping

**Review by russoz — APPROVED** [review]
LGTM

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:12]: Who is @m-a-r-k-e and how were they involved in this? Being added here and in the maintainers file means that they will get pinged for every issue / PR filed for this plugin. Is that OK?

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [inline plugins/lookup/merge_variables.py:18]: Note to self: this needs to be adjusted before merging.

File: plugins/lookup/merge_variables.py
Comment (m-a-r-k-e) [reply plugins/lookup/merge_variables.py:12]: @rlenferink and I developed this plugin together. I was already following this, so it is OK to get pinged for any future issues/questions.

Comment (felixfontein) [reply plugins/lookup/merge_variables.py:12]: Thanks for confirming that here. I don't like adding users who weren't involved (and where it is unclear whether they consent to it) :)

File: plugins/lookup/merge_variables.py
Comment (felixfontein) [reply plugins/lookup/merge_variables.py:18]:
```suggestion
    version_added: 6.5.0
```

**Comment by felixfontein** [comment]
@rlenferink @m-a-r-k-e thanks for your contribution!
@russoz thanks for reviewing!

**Comment by rlenferink** [comment]
@felixfontein @russoz thanks for reviewing !!

@indelog let me know if you want to progress with #4799. I am happy to think along and see whether parts can be combined in the same module.

**Pull request merged**
